---
title: "KotggenDataPrep"
author: "Nick Burns"
date: "8 April 2016"
output: html_document
---

# Kottgen vs. Leffler  

The plan is to identify regions of interest in the Leffler dataset, and to create classical locus zooms of these same regions in the Kottgen data set. To do this we will need to prepare the Kottgen dataset (first section beneath), identify regions of interest in Leffler (second section) and finally visualise (thrid section).


## Data prep for the loci exploration of the Kottgen dataset.

Kottgen summary level statistics were downloaded from: http://metabolomics.helmholtz-muenchen.de/gugc/, 8 April 2015. The summary level files are apprx. 130 MB each, which isn't huge, but we don't need ot deal with quite so much data.

Filter datasets to P <= 0.1.


```{r}
setwd("/mnt/DataDrive/Kottgen")
library(data.table)

urate <- fread("GUGC_MetaAnalysis_Results_UA.csv", header=TRUE, data.table = FALSE)
dim(urate)

gout <- fread("GUGC_MetaAnalysis_Results_Gout.csv", header=TRUE, data.table = FALSE)
dim(gout)
```

Here is hte head of these two files:

```{r}
head(urate)
head(gout)
```

Looking at the distribution of p-values:

```{r}
hist(urate$p_gc)
hist(gout$p_gc)
```

And we will filter these to p-values <= 0.01, which we can see from the above plots will end up being < 10 % the size.

```{r}
filterGwas <- function (data, threshold = 0.01) {
    return (data[data$p_gc < threshold, ])
}

write.csv(filterGwas(urate), "Kottgen_Urate_SummaryStats_Filtered.csv", row.names=FALSE)
write.csv(filterGwas(gout), "Kottgen_Gout_SummaryStats_Filtered.csv", row.names=FALSE)
```

### Get chromosome and position  

These summary files do not include chromosome numbers or position. Need to query these based on the marker name. I will test this function here, and then include it in GLIDA.

```{r}
urateSet <- filterGwas(urate, threshold=0.1)
goutSet <- filterGwas(gout, threshold = 0.1)

library(RMySQL)

queryPos <- function (snps, conn) {
    
    formatSNPs <- sprintf("('%s')", paste(snps, collapse="','"))
    query <- paste0("
                    SELECT chrom,
                           chromStart as POS,
                           name as MarkerName,
                           strand
                    FROM snp142
                    WHERE name IN ", formatSNPs, "
                      AND chrom NOT LIKE 'chr%hap%';
                    ")
    
    results <- dbGetQuery(conn, query)
    
    return (results)
}

updatePos <- function (snpList) {
    
    options(warn = -1)
    conn <- dbConnect(RMySQL::MySQL(),
                           user = "genome", 
                           host = "genome-mysql.cse.ucsc.edu",
                           dbname = "hg19", 
                           password = "")
    
    snpPositions <- queryPos(snpList$MarkerName, conn)
    
    snpList <- merge(snpList, snpPositions, by="MarkerName")
    
    dbDisconnect(conn)
    options(warn = 0)
    
    return (snpList)
}

urateUpdated <- updatePos(urateSet)
goutUpdated <- updatePos(goutSet)
```

Parse the chromosome column and remove the 'chr' prefix:

```{r}
urateUpdated$chrom <- sub("chr", "", urateUpdated$chrom)
goutUpdated$chrom <- sub("chr", "", goutUpdated$chrom)

head(urateUpdated)
head(goutUpdated)

write.csv(urateUpdated, "Kottgen_Urate_SummaryStats_Filtered.csv")
write.csv(goutUpdated, "Kottgen_Gout_SummaryStats_Filtered.csv")
```

## Leffler: Identify Regions of interest  

Supplmentary tables 4 and 5 are a summary of interesting SNPs that were identified in the Leffler paper. We will cluster these to try to reduce the total number of regions we need to plot.

```{r}
setwd("/home/nickb/Documents/GitHub/Kottgen_ReseqAnalysis")

kottgenUrate <- read.csv("Data/Kottgen_Urate_SummaryStats_Filtered.csv")
leffler4 <- read.csv("Data/LefflerTablesS4.csv")
leffler5 <- read.csv("Data/LefflerTablesS5.csv")

head(leffler4)
head(leffler5)
```

We will combine both of these tables and treat them as a single dataset to explore.

```{r}
lefflerCombined <- leffler4[, c("Chr19", "Position_hg19", "SNP")]
colnames(lefflerCombined) <- c("CHR", "POS", "SNP")
lefflerCombined <- rbind(lefflerCombined,
                         data.frame(CHR = leffler5$Chr_hg19, 
                                    POS = leffler5$Position_hg19,
                                    SNP = leffler5$SNP))
dim(lefflerCombined)
head(lefflerCombined)
```

Use DBSCAN for clustering  

```{r}
library(dbscan)
getWindows <- function (data, eps=0.5, minPts=1) {
    
    # Finds windows of interest by clustering the SNPs in each chromosome.
    #
    # Parameters:
    # -----------
    #
    #     data: data frame (CHR, SNP, POS)
    #     eps: float
    #         Neighbourhood radius for DBSCAN
    #         DEFAULT = 0.5 (500KB)
    #     minPts: int
    #         ncore points parameter for DBSCAN
    #         DEFAULT = 1
    #
    # Returns:
    # --------
    #
    #     data: data frame (CHR, SNP, POS, P, Phenotype, Nation, LociID)
    #         GWAS Summary data.
    
    data$LociID <- 0  # initially set all SNPs to a NULL loci / window
    
    # cluster by chromosome
    # NOTE: by chr so that we don't end up with windows which span chr.
    for (chrom in unique(data$CHR)) {
        
        idxCHR <- data$CHR == chrom
        window <- dbscan(as.matrix(data[idxCHR, "POS"])/1000000, 
                         eps=eps, minPts=minPts)
        
        # record the LociID (i.e. window ID) for all SNPs
        # NOTE: DBSCAN always returns {1, 2, ...}, so we need to offset
        # these integer labels by the current max to ensure we don't repeat IDs
        data$LociID[idxCHR] <- window$cluster + max(data$LociID)
        
    }
    return (data)
}
leffler <- getWindows(lefflerCombined, eps=2)
leffler <- unique(leffler)
```

Nice, that was super quick. If we inspect these:

```{r}
nrow(lefflerCombined)
max(leffler$LociID)
```

Ouch, even with a neighbourhood radius of 2 MB, there are still 294 loci. That is too many to be realistic. I need to think about this a little.

## Kottgen Loci

294 loci are too many really. I need to think of ways to summarise the characteristics of these loci. If I can describe the Kottgen data at each loci,then I might be able to find a way to reduce complexity and produce a sensible result set. The first job then is to extract these loci from the Kottgen data.

For each leffler loci, I will extract the CHR, min(POS) and max(POS) and use these to extract the same regions from Kottgen.

```{r}
library(dplyr)
loci <- dplyr::summarise(group_by(leffler, LociID),
                         CHR = unique(CHR),
                         start = min(POS),
                         end = max(POS),
                         nSNPs = n())
head(loci)
```

Let's make life really easy to start with and plot the first few of these. First, we will extract a trial dataset and then plot it:

```{r}
extractKottgen <- function (kottgenData, region, delta = 500000) {
    
    lclChr <- region$CHR
    lclStart <- region$start - delta
    lclEnd <- region$end + delta

    #print(c(lclChr, lclStart, lclEnd))
    loci <- kottgenData[(kottgenData$chrom == lclChr) & 
                        (kottgenData$POS >= lclStart) &
                            (kottgenData$POS <= lclEnd), ]
    if (nrow(loci) > 0)
        loci$LociID <- unique(region$LociID)
    
    return (loci)
}

kottgenLoci <- do.call(rbind, 
                     lapply(unique(loci$LociID), 
                            function (k) extractKottgen(kottgenUrate, loci[loci$LociID == k, ])))
dim(kottgenLoci)
head(kottgenLoci)
```


### Summary stats of the Kottgen Loci

Given that there are close to 300 loci, let's look at some summary stats for these loci.

1. Loci by chromosome  

```{r}
library(ggplot2)

ggplot(kottgenLoci, aes(x = chrom, y = LociID)) + 
    geom_point(colour="steelblue") + 
    theme_bw()
```

Not overly surprising here, that there seem to be many loci per chromosome, with more loci on the larger chromosomes. Let's consider the same plot but where each loci is sized by the p-value.

```{r}
library(ggrepel)
kottgenByLoci <- summarise(group_by(kottgenLoci, chrom, LociID),
                           nLoci = n(),
                           maxP = max(-log10(p_gc)))
head(kottgenByLoci)

ggplot(kottgenByLoci, aes(x = chrom, y = LociID)) +
    geom_point(aes(size = maxP), colour = "steelblue", alpha = 0.7) +
    theme_bw() +
    geom_text_repel(data = kottgenByLoci[kottgenByLoci$maxP > 5, ],
              aes(x = chrom, y = LociID, label = LociID),
              size = 5)
```

That's more interesting. Loci with p-value < 10-5 are labeled above. This gives us a much smaller number to look at. Let's specifically look at these loci:

```{r}
targetLoci <- c(77, 109, 133, 139, 157, 192, 213, 258, 260, 267)
ggplot(kottgenLoci[kottgenLoci$LociID %in% targetLoci, ],
       aes(x = POS / 1000000, y = -log10(p_gc))) +
    geom_point(colour = "grey") +
    facet_wrap(~chrom, scales = "free_x") +
    theme_bw() +
    geom_hline(yintercept = 5, colour = "blue", linetype = "dashed") +
    geom_hline(yintercept = 8, colour = "red", linetype = "dashed") +
    xlab('POS') + ylab("-log10(P)")
```

This is all good now. Clearly there are some interesting areas.

### Plots for Tony  

For Tony, I will plot all of the Loci and send these to him. However, I will order them by p-value.

```{r}
library(gridExtra)

plots <- list()
for (lx in seq(length(targetLoci))) {
    
    dx <- kottgenLoci[kottgenLoci$LociID == lx, ]
    dx$position <- dx$POS / 1000000
    
    
    zoom <- ggplot(dx, aes(x = position, y = -log10(p_gc))) +
        geom_point() +
        theme_bw()
    
    genes <- queryUCSC(fromUCSCEnsemblGenes(chromosome = unique(dx$chrom),
                                            start = min(dx$POS),
                                            end = max(dx$POS)))
    
    zoom <- geneAnnotation(zoom, genes)
    plots[[lx]] <- zoom
     
}


ggsave("testplots.pdf", do.call(arrangeGrob, plots))

```

